File: av1an/chunk/Chunk.py
from pathlib import Path
from typing import Dict, Any


class Chunk:
    def __init__(
        self,
        temp: Path,
        index: int,
        ffmpeg_gen_cmd: list,
        output_ext: str,
        size: int,
        frames: int,
    ):
        self.index: int = index
        self.ffmpeg_gen_cmd: str = ffmpeg_gen_cmd
        self.size: int = size
        self.temp: Path = temp
        self.frames: int = frames
        self.output_ext: str = output_ext
        self.per_shot_target_quality_cq = None

    def to_dict(self) -> Dict[str, Any]:

        return {
            "index": self.index,
            "ffmpeg_gen_cmd": self.ffmpeg_gen_cmd,
            "size": self.size,
            "frames": self.frames,
            "output_ext": self.output_ext,
            "per_shot_target_quality_cq": self.per_shot_target_quality_cq,
        }

    @property
    def output_path(self) -> Path:
        return (self.temp / "encode") / f"{self.name}.{self.output_ext}"

    @property
    def output(self) -> str:

        return self.output_path.as_posix()

    @property
    def name(self) -> str:

        return str(self.index).zfill(5)

    @staticmethod
    def create_from_dict(d: dict, temp):
        chunk = Chunk(
            temp,
            d["index"],
            d["ffmpeg_gen_cmd"],
            d["output_ext"],
            d["size"],
            d["frames"],
        )
        chunk.per_shot_target_quality_cq = d["per_shot_target_quality_cq"]
        return chunk

File: av1an-core/src/chunk.rs
#[cfg(test)]
mod tests;

use std::{ffi::OsString, path::Path};

use av1_grain::{generate_photon_noise_params, write_grain_table, NoiseGenArgs};
use serde::{Deserialize, Serialize};
use tracing::debug;

use crate::{encoder::Encoder, settings::insert_noise_table_params, Input};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Chunk {
    pub temp:                  String,
    pub index:                 usize,
    pub input:                 Input,
    pub source_cmd:            Vec<OsString>,
    pub output_ext:            String,
    pub start_frame:           usize,
    // End frame is exclusive, i.e. the range of frames is `start_frame..end_frame`
    pub end_frame:             usize,
    pub frame_rate:            f64,
    pub passes:                u8,
    pub video_params:          Vec<String>,
    pub encoder:               Encoder,
    pub noise_size:            (Option<u32>, Option<u32>),
    // do not break compatibility with output produced by older versions of av1an
    /// Optional target quality CQ level
    #[serde(rename = "per_shot_target_quality_cq")]
    pub tq_cq:                 Option<u32>,
    pub ignore_frame_mismatch: bool,
}

impl Chunk {
    /// Returns numeric name of chunk `00001`
    pub fn name(&self) -> String {
        format!("{:05}", self.index)
    }

    pub fn output(&self) -> String {
        Path::new(&self.temp)
            .join("encode")
            .join(format!("{}.{}", self.name(), self.output_ext))
            .to_str()
            .unwrap()
            .to_owned()
    }

    pub const fn frames(&self) -> usize {
        self.end_frame - self.start_frame
    }

    pub(crate) fn apply_photon_noise_args(
        &mut self,
        photon_noise: Option<u8>,
        chroma_noise: bool,
    ) -> anyhow::Result<()> {
        if let Some(strength) = photon_noise {
            let iso_setting = u32::from(strength) * 100;
            let grain_table = Path::new(&self.temp).join(format!("iso{iso_setting}-grain.tbl"));
            if !grain_table.exists() {
                debug!("Generating grain table at ISO {iso_setting}");
                let (mut width, mut height) = self.input.resolution()?;
                if self.noise_size.0.is_some() {
                    width = self.noise_size.0.unwrap();
                }
                if self.noise_size.1.is_some() {
                    height = self.noise_size.1.unwrap();
                }
                let transfer_function =
                    self.input.transfer_function_params_adjusted(&self.video_params)?;
                let params = generate_photon_noise_params(0, u64::MAX, NoiseGenArgs {
                    iso_setting,
                    width,
                    height,
                    transfer_function,
                    chroma_grain: chroma_noise,
                    random_seed: None,
                });
                write_grain_table(&grain_table, &[params])?;
            }

            insert_noise_table_params(self.encoder, &mut self.video_params, &grain_table)?;
        }

        Ok(())
    }
}
